### PR TITLE
Support instrument field collector

### PR DESCRIPTION
**Pull request by dugenkui03**
Details about this PR #2525 .

### CODE REVIEW

File: src/main/java/graphql/execution/ExecutionStrategy.java [hunk 1]
```diff
@@ -648,6 +648,8 @@ protected FieldValueInfo completeValueForList(ExecutionContext executionContext,
                 .objectType(resolvedObjectType)
                 .fragments(executionContext.getFragmentsByName())
                 .variables(executionContext.getVariables())
+                .instrumentation(executionContext.getInstrumentation())
+                .instrumentationState(executionContext.getInstrumentationState())
```
Comment (dugenkui03) [inline src/main/java/graphql/execution/ExecutionStrategy.java:652]: For use at `FieldCollector`.

File: src/main/java/graphql/execution/FieldCollector.java [hunk 2]
```diff
@@ -38,7 +40,16 @@ public MergedSelectionSet collectFields(FieldCollectorParameters parameters, Mer
...
+
+            Map<String, MergedField> fields = new LinkedHashMap<>();
+            this.collectFields(parameters, field.getSelectionSet(), visitedFragments, fields);
+
+            InstrumentationCollectFieldsParameters collectFieldsParameters = new InstrumentationCollectFieldsParameters(
+                    parameters, field.getSelectionSet(), visitedFragments, fields, parameters.getInstrumentationState()
+            );
+
+            fields = parameters.getInstrumentation().instrumentFieldsCollect(collectFieldsParameters, fields);
+            subFields.putAll(fields);
```
Comment (dugenkui03) [inline src/main/java/graphql/execution/FieldCollector.java:52]: Alter the collect behavior by `instrumentFieldsCollect `.

File: src/main/java/graphql/execution/FieldCollectorParameters.java [hunk 3]
```diff
@@ -17,6 +19,8 @@
     private final Map<String, FragmentDefinition> fragmentsByName;
     private final Map<String, Object> variables;
     private final GraphQLObjectType objectType;
+    private final Instrumentation instrumentation;
+    private final InstrumentationState instrumentationState;
```
Comment (dugenkui03) [inline src/main/java/graphql/execution/FieldCollectorParameters.java:23]: For alter the collect behavior at `FieldCollector`.

File: src/main/java/graphql/execution/instrumentation/parameters/InstrumentationCollectFieldsParameters.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import graphql.execution.MergedField;
+import graphql.execution.instrumentation.InstrumentationState;
+import graphql.language.SelectionSet;
+
+import java.util.Map;
+import java.util.Set;
+
+@PublicApi
+public class InstrumentationCollectFieldsParameters {
+    private final FieldCollectorParameters fieldCollectorParameters;
```
Comment (dugenkui03) [inline src/main/java/graphql/execution/instrumentation/parameters/InstrumentationCollectFieldsParameters.java:15]: The information maybe used at `instrumentFieldsCollect`.

File: src/main/java/graphql/execution/nextgen/FieldCollector.java [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+import static graphql.execution.MergedSelectionSet.newMergedSelectionSet;
+import static graphql.execution.TypeFromAST.getTypeFromAST;
+/**
+ * A field collector can iterate over field selection sets and build out the sub fields that have been selected,
+ * expanding named and inline fragments as it goes.s
+ */
+@Internal
+public class FieldCollector {
+
```
Comment (dugenkui03) [inline src/main/java/graphql/execution/nextgen/FieldCollector.java:33]: Copy old code, in order to don't effect `nextgen` behavior.

File: src/main/java/graphql/execution/nextgen/FieldCollectorParameters.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+import graphql.schema.GraphQLSchema;
+
+import java.util.Map;
+
+/**
+ * Internal because FieldCollector is internal.
+ */
+@Internal
+public class FieldCollectorParameters {
+    private final GraphQLSchema graphQLSchema;
```
Comment (dugenkui03) [inline src/main/java/graphql/execution/nextgen/FieldCollectorParameters.java:16]: Copy old code, in order to don't effect `nextgen` behavior.

File: src/test/groovy/graphql/execution/FieldCollectorTest.groovy [hunk 7]
```diff
@@ -28,6 +34,8 @@ class FieldCollectorTest extends Specification {
         FieldCollectorParameters fieldCollectorParameters = newParameters()
                 .schema(schema)
                 .objectType(objectType)
+                .instrumentation(SimpleInstrumentation.INSTANCE)
+                .instrumentationState(new CustomInstrumentationState())
```
Comment (dugenkui03) [inline src/test/groovy/graphql/execution/FieldCollectorTest.groovy:38]: Avoid exception caused by not set `instrumentation`. `instrumentFieldsCollect` will be invoked in this new code.

File: src/main/java/graphql/execution/nextgen/ExecutionHelper.java [hunk 8]
```diff
@@ -5,8 +5,6 @@
 import graphql.execution.ExecutionContext;
 import graphql.execution.ExecutionId;
 import graphql.execution.ExecutionStepInfo;
-import graphql.execution.FieldCollector;
-import graphql.execution.FieldCollectorParameters;
 import graphql.execution.MergedSelectionSet;
```
Comment (dugenkui03) [inline src/main/java/graphql/execution/nextgen/ExecutionHelper.java:8]: Use `graphql.execution.nextgen.FieldCollector` and `graphql.execution.nextgen.FieldCollectorParameters` which keep same before this pr.

File: src/main/java/graphql/execution/nextgen/ExecutionStrategyUtil.java [hunk 9]
```diff
@@ -6,8 +6,6 @@
 import graphql.execution.ExecutionStepInfo;
 import graphql.execution.ExecutionStepInfoFactory;
 import graphql.execution.FetchedValue;
-import graphql.execution.FieldCollector;
-import graphql.execution.FieldCollectorParameters;
 import graphql.execution.MergedField;
```
Comment (dugenkui03) [inline src/main/java/graphql/execution/nextgen/ExecutionStrategyUtil.java:9]: Use `graphql.execution.nextgen.FieldCollector` and `graphql.execution.nextgen.FieldCollectorParameters` which keep same before this pr.

**Comment by bbakerman** [comment]
I am not sure we would accept this PR.  

Collecting the fields of a query has proven to be one of the most challenging aspects of building a graphql engine.

We have recently invested in a bunch of code around field collection via the new `graphql.normalized.ExecutableNormalizedField` classes and the main implementation `graphql.normalized.ExecutableNormalizedOperationFactory`

These classes now back our "field selection code" in the DataFecther plus we are likely to make it public API in some shape in future releases.

So field collection is a non trivial problem.  

So if we allow an SPI to "change" the behavior of "field collection" then it makes it very hard for us to have this other code.

We really want a canonical algorithm for field collection - one that does not change and hence the execution engine can be written in terms of it.

What you have here is really a new "execution engine".  You are calling out to decide that fields are included or not.   I think rather than instrumentation this is a "new engine".

These are my thoughts, @andimarek may have more thoughts on this.

**Comment by haggaikaunda** [comment]
Hey folks, Haggai here from Twitter.

Weighing on this discussion, we have a very similar use case at Twitter that would be solved by this functionality. This is similar to @skip/@include directives which are handled by `ConditionalNodes` class in `FieldCollector.java`

The user story in our case is that Client engineers sometimes need to use the combination of @skip/@include directives to conditionally fetch nodes in cases of migration, new feature, experient etc. As you might imagine, those directives would be driven by some sort of feature variable that can be toggled on and off depending on the state of the experiment for example.

Now because graphql requires global variables be defined on query definition, this means that if such a feature variable is referenced on an important fragment such as a Tweet which, at Twitter, is used to compose 100s of queries, an engineer needs to go and manually update ALL the query definitions that reference the said fragment. This process is painful and error prone. We are actively looking to eliminate the need to passing these feature variables to the query and have their values supplied else where at execution time. A concrete pseudo example, we want to convert something like this...

```
fragment Tweet on Tweet {
  text
  experimentalFeature @include(if:$enableExperiment)
}

query HomeTimeline($enableExperiment) {
  homeTimeline {
    ...Tweet
  }
}

query ProfileTimeline($enableExperiment) {...}

# PS: There's literally 100s of queries that need to be manually updated everytime we introduce a variable like this
```

into something that removes the need to declare `enableExperiement` as a variable on each of the queries definitions that reference the Tweet fragment, we can do this by

```
fragment Tweet on Tweet {
  text
  experimentalFeature @experimentFlag(exp:"experimentFeatureFlag")
}

query HomeTimeline {
  homeTimeline {
    ...Tweet
  }
}

query ProfileTimeline {...} # 100s of queries referencing Tweet fragment don't need manual updating!
```

The ability to modify the selection set at execution time allows us to create this directive, and thus eliminating the need for complicated variable writing by our client developers.

PS: I also found a similar post on [spectrum](https://spectrum.chat/graphql-java/general/how-to-write-custom-directive-that-omits-result-fields-like-skip~85b2048e-731a-4a15-82de-16be606c8757) from 2020 about adding such functionality to graphql java.

**Comment by bbakerman** [comment]
@haggaikaunda 

In your case I think you can do what you want via query rewriting, rather than present the query to the engine and then use instrumentation to change what fields get collected.

See `graphql.analysis.QueryTransformer`

```
        def query = TestUtil.parseQuery("{ root { fooA { midA { leafA } midB { leafB } } fooB { midA { leafA } midB { leafB } } } }")

        QueryTransformer queryTransformer = createQueryTransformer(query, schema)

        def visitor = new QueryVisitorStub() {
            @Override
            void visitField(QueryVisitorFieldEnvironment env) {
                if (env.fieldDefinition.type.name == "MidA") {
                    deleteNode(env.getTraverserContext())
                }
            }
        }


    QueryTransformer createQueryTransformer(Document document, GraphQLSchema schema, Map variables = [:]) {
        def fragments = NodeUtil.getFragmentsByName(document)
        QueryTransformer queryTransformer = QueryTransformer.newQueryTransformer()
                .schema(schema)
                .fragmentsByName(fragments)
                .root(document)
                .variables(variables)
                .rootParentType(schema.getQueryType())
                .build()
        return queryTransformer
    }

```

(groovy code from our tests)

This would allow you to take a query and filter out fields from it and create a new Document.

you can do this in

`graphql.execution.instrumentation.Instrumentation#instrumentDocumentAndVariables`

This is called after the original query is parsed into a Document but before its validated.

**Comment by haggaikaunda** [comment]
Hey @bbakerman, 

We actually already have a prototype that does exactly that. We are using`PreparsedDocumentProvider#getDocument`, which is the method that wraps the execution of `Instrumentation#instrumentDocumentAndVariables` and have logic to rewrite the query at execution time. The call to `instrumentDocumentAndVariables` method is wrapped in a call-back function (`parseAndValidateFunction`) that only gets executed when a query has not been pre-parsed. The call back, along with invoking the instrumentation will parse and validate the query. 

Meaning that rewriting requires that we parse and validate queries on every request - this creates other issues where caching preparsed queries is not as straightforward. If you want to cache a query with such a directive, the cache size grows exponentially based on the number of directives (@experimentalFlag).

**Comment by bbakerman** [comment]
Again let me stress why we are against this PR.  If we allow Instrumentation to change how fields are collected at runtime it affects a bunch of code beyond that instrumentation.

We are planning to look into areas where we build the list of fields to be executed before we get to the execution engine and also we want it to be "algorithmically stable" eg some code can assume that the fields are the same according to a well known algorithm

We feel the dynamic nature of Instrumentation means we will always be second guessing things ongoing. eg this is the set if fields except if the instrumentation changes them.

> Meaning that rewriting requires that we parse and validate queries on every request - this creates other issues where caching preparsed queries is not as straightforward. 

in this case you would be better to pre-parse and change the query text before sending it to the engine at all - so rather than a call back, change it before it is ever presented.

**Comment by haggaikaunda** [comment]
Thanks for the consideration and explanation @bbakerman. We'll look at other alternatives such as transforming the @experimentalFlag directives into equivalent @skip/@include in a way that makes the query reusable across multiple requests.

**Pull request closed without merging**